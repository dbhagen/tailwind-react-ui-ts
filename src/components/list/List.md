### Simple Example

```jsx
<List padding ordered list>
  <Text>One</Text>
  <Text>Two</Text>
  <Text>Three</Text>
</List>
```

### Inline

```jsx
<List padding inline>
  <Text>One</Text>
  <Text>Two</Text>
  <Text>Three</Text>
</List>
```

### Justified

```jsx
<List justified>
  <Text>One</Text>
  <Text>Two</Text>
  <Text>Three</Text>
</List>
```

### Full Width

```jsx
<List fullWidth>
  <Text text="center" p={2} bg="gray-200" block>
    One
  </Text>
  <Text text="center" p={2} bg="gray-300" block>
    Two
  </Text>
  <Text text="center" p={2} bg="gray-200" block>
    Three
  </Text>
</List>
```
