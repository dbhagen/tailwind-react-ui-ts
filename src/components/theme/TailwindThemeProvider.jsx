import React from 'react'
import PropTypes from 'prop-types'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

const TailwindThemeProvider = ({ theme, children }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

TailwindThemeProvider.propTypes = {
  theme: PropTypes.shape({}),
  children: PropTypes.node,
}

TailwindThemeProvider.defaultProps = {
  theme: {},
  children: undefined,
}

export default TailwindThemeProvider
