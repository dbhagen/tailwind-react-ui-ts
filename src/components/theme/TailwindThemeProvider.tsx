import React from 'react'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

export type TailwindThemeProviderProps = {
  theme: {}
  children: React.ReactNode
}

const TailwindThemeProvider: React.FunctionComponent<TailwindThemeProviderProps> = ({ theme, children }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

export default TailwindThemeProvider
