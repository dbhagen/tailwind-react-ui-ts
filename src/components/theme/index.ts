export { default as defaultTheme } from './defaultTheme'
export { default as TailwindTheme } from './TailwindTheme'
export { default as TailwindThemeProvider } from './TailwindThemeProvider'
export { default as withTheme } from './withTheme'
export { default as useThemeValue } from './useThemeValue'
