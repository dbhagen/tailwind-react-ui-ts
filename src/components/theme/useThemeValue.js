import { tailwindPropToClassName } from '../tailwind'

export default (prefix, value, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
