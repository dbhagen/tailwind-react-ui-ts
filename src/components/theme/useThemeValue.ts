import { tailwindPropToClassName } from '../tailwind'

export default (prefix: string, value: string, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
