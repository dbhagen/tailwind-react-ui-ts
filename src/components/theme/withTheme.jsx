import React from 'react'

import TailwindTheme from './TailwindTheme'

const withTheme = Component => {
  const WithTheme = props => (
    <TailwindTheme.Consumer>
      {theme => <Component {...props} theme={theme} />}
    </TailwindTheme.Consumer>
  )

  WithTheme.displayName = `WithTheme(${Component.displayName})`

  return WithTheme
}

export default withTheme
