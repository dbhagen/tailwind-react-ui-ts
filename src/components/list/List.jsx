import React from 'react'
import PropTypes from 'prop-types'

import { Box } from '../primitives'
import { withTheme } from '../theme'

const List = ({
  theme,
  is,
  children,
  padding,
  list,
  inline,
  justified,
  fullWidth,
  ordered,
  listItemIs,
  ...rest
}) => {
  let listStyle = list
  if (listStyle === true) {
    listStyle = ordered ? 'decimal' : 'disc'
  }

  return (
    <Box
      is={ordered ? 'ol' : is}
      m={{ b: theme.spacing.md }}
      p={
        listStyle
          ? {
            l: theme.spacing.md
          }
          : undefined
      }
      flex={
        justified ||
        fullWidth ||
        (inline ? [true, 'wrap', ...rest.flex] : rest.flex)
      }
      justify={justified ? 'between' : undefined}
      list={listStyle}
      {...rest}
    >
      {React.Children.map(
        children,
        child =>
          child && (
            <Box
              is={listItemIs}
              m={{
                b: padding && !justified && !fullWidth && theme.spacing.sm,
                r: inline && theme.spacing.sm
              }}
              flex={fullWidth ? 'grow' : undefined}
            >
              {child}
            </Box>
          )
      )}
    </Box>
  )
}

List.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  padding: PropTypes.bool,
  list: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.oneOf(['disc', 'decimal'])
  ]),
  inline: PropTypes.bool,
  justified: PropTypes.bool,
  fullWidth: PropTypes.bool,
  ordered: PropTypes.bool,
  listItemIs: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
    PropTypes.object
  ])
}

List.defaultProps = {
  is: 'ul',
  children: undefined,
  padding: false,
  list: undefined,
  inline: false,
  justified: false,
  fullWidth: false,
  ordered: false,
  listItemIs: 'li'
}

export { List as component }
export default withTheme(List)
