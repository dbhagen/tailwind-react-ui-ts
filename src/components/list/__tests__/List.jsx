import React from 'react'
import { shallow } from 'enzyme'

import { component as List } from '../List'
import { defaultTheme } from '../../theme'

const setup = (testProps = {}) => {
  const props = Object.assign({ theme: defaultTheme }, testProps)

  const wrapper = shallow(
    <List inline {...props}>
      <span>One</span>
      <span>Two</span>
      <span>Three</span>
    </List>,
  )

  return {
    props,
    wrapper,
  }
}

describe('List', () => {
  it('renders matching snapshot', () => {
    const { wrapper } = setup()

    expect(wrapper).toMatchSnapshot()
  })
})
