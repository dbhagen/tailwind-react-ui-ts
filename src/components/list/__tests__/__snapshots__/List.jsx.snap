// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders matching snapshot 1`] = `
<Box
  flex={
    Array [
      true,
      "wrap",
      undefined,
    ]
  }
  inline={false}
  inlineBlock={false}
  is="ul"
  m={
    Object {
      "b": 4,
    }
  }
>
  <Box
    inline={false}
    inlineBlock={false}
    is="li"
    key=".0"
    m={
      Object {
        "b": false,
        "r": 2,
      }
    }
  >
    <span>
      One
    </span>
  </Box>
  <Box
    inline={false}
    inlineBlock={false}
    is="li"
    key=".1"
    m={
      Object {
        "b": false,
        "r": 2,
      }
    }
  >
    <span>
      Two
    </span>
  </Box>
  <Box
    inline={false}
    inlineBlock={false}
    is="li"
    key=".2"
    m={
      Object {
        "b": false,
        "r": 2,
      }
    }
  >
    <span>
      Three
    </span>
  </Box>
</Box>
`;
